student_work/julesfowler_work/reduce_pluto.py

```diff
@@ -0,0 +1,222 @@
+## -- IMPORTS
+
+import glob
+
+from astropy.io import fits
+from astropy.stats import sigma_clip
+import matplotlib.pyplot as plt
+import numpy as np
+from photutils import centroid_1dg, centroid_2dg
+
+
+## -- FUNCTIONS
+
+def calibrate_data(biases='./biases/d*.fits', darks='./darks/night/d*.fits', 
+                   flats='./dome_flats/d*fits', science='./science/d*.fits',
+                   writeout='final_pluto.fits', plot='final_pluto.png'):
+    """ Main function to calibrate data from start to finish. 
+    Note that with the default paths this is meant to run in a particuarly
+    data directory structure. Future iterations will probably change that.
+
+    Parameters
+    ----------
+    biases: str
+        Glob-able path to bias fits files.
+    darks: str
+        Glob-able path to dark fits files.
+    flats: str
+        Glob-able path to flat fits files. (Will work with sky or dome flats.
+    science: str
+        Glob-able path to science fits files.
+    writeout: str
+        Name to write out the final calibrated pluto science image to.
+    plot: str
+        Name to write out the final calibrated pluto plot to.
+    
+    Returns
+    -------
+    science_image : np.array
+        Array of reduced science data.
+    """
+
+    bias_image = make_bias(biases)
+    dark_image = make_dark(darks, bias_image, writeout='dark.fits')
+    flat_image = make_flat(flats, dark_image, writeout='flat.fits')
+    science_image = make_science_image(science, bias_image, dark_image, 
+            flat_image, writeout=writeout, plot=plot)
+
+    return science_image
+
+
+def make_bias(bias_path):
+    """ Quick one-liner to return the super bias in a modular way."""
+    return median_combine(bias_path)
+
+def make_dark(dark_path, bias_image, writeout=None):
+    """ Builds the bias subtracted dark image.
+    
+    Parameters
+    ----------
+    dark_path : str
+        Glob-able path to dark fits files.
+    bias_image : np.array
+        Super bias in numpy array form.
+    writeout : str or None
+        Name of the fits file to write out, if at all.
+    
+    Returns
+    -------
+    super_dark_per_second : np.array
+        Bias subtracted + time averaged super dark.
+    """
+
+    super_dark = median_combine(dark_path)
```

Comment:
are all of the darks restricted to have the same exposure time?
If not, this seems problematic.

Reply:
Yup! All darks are taken with the same exposure time.

Reply:
If I were being very rigorous I could check each exposure time and throw an error if they aren't...